scan-external-contacts/nodejs/index.js

```diff
@@ -0,0 +1,296 @@
+
+        return false;
+    }
+}
+
+// Perform an External Contacts Scan API request and manage errors (429)
+const scanContacts = async (cursor) => {
+
+    // Defining the default values for the result structure
+    let result = {
+        error: false,
+        errorMsg: "",
+        entities: [],
+        isEmpty: true,
+        isFinished: true,
+        nextCursor: null,
+        startTime: 0
+    };
+
+    try {
+        result.startTime = Date.now();
+
+        let opts = {
+            'limit': maxRecordsPerPage
+        };
+        // No opts.cursor on first scan - cursor == null
+        if (cursor) {
+            opts.cursor = cursor;
+        }
+
+        // External Contacts Scan API request
+        let scanData = await contactsApi.getExternalcontactsScanContacts(opts);
+        // If the request is successful, we analyze the response
+
+        // The response may contain an empty "entities" array in certain conditions (congestion on database, contacts deleted before the request was performed, ...)
+        if (scanData.entities && scanData.entities.length > 0) {
+            result.isEmpty = false;
+            result.entities = scanData.entities;
+        }
+
+        // If the response contains an "after" attribute inside "cursors" (cursors.after), then, there are still some External Contacts (data) to retrieve
+        // When the response does not contain an "after" attribute inside "cursors", it means it is the last page of scan
+        if (scanData.hasOwnProperty("cursors") && scanData.cursors.hasOwnProperty("after")) {
+            result.isFinished = false;
+            result.nextCursor = scanData.cursors.after;
+        }
+
+        // Reset of some global vars - counters for retry
+        consecutiveRateLimitRetryAttempts = 0;
+
+        return result;
+    }
+    catch (error) {
+        console.log('External Contacts Scan error - Error:', JSON.stringify(error, null, 2));
+
+        if (error.status && error.status === 429) {
+            // API Rate Limit - Max number of API requests per minute reached
+            // Exit Application if too many consecutive retry failures
+            if (consecutiveRateLimitRetryAttempts >= maxRateLimitRetryAttempts) {
+                result.errorMsg = "Max consecutive API Rate Limit. Exiting...";
+                result.error = true;
+                return result;
+            } else {
+                // Extract retry-after header value (seconds) from the 429 error
+                let retryAfter = 0;
+                if (error.headers && error.headers.hasOwnProperty("retry-after")) {
+                    retryAfter = parseInt(error.headers["retry-after"]);
+                } else {
+                    result.errorMsg = "Failed to retrieve Retry-After. Exiting...";
+                    result.error = true;
+                    return result;
+                }
+                // Add 1 second (for "safety")
+                let retryIn = (retryAfter + 1) * 1000;
```

Comment:
Rather than writing this yourself, you could steer the reader to using a library like lifeomic's attempt library.  I used it in the user-provisioning-scripts developer guide and might be a better solution than rolling our own.  (https://github.com/lifeomic/attempt)